[mono][wasm] Bundle assemblies as WebCIL

lambdageek opened the pull request:
Define a new container format for .NET assemblies that looks less like a Windows PE file.  Use it for bundling assemblies in wasm projects.

TODO:

- [x] Turn off Mono support for webcil on non-WebAssembly platforms
- [x] Turn webcil bundling off by default. Add an MSBuild property that users can set to preview webcil bundling. **done.  Set `<WasmEnableWebcil>true</WasmEnableWebcil>` to try it
- [x] Run Wasm.Build.Test tests 2x: with and without webcil



Nice to have:
* Maybe do something with InternalsVisibleTo so that we can use the internal System.Reflection.Metadata classes to implement WebcilReader and WebcilConverter.  At least on netcore.  (on .NET Framework we must depend on the 6.0.0 SRM because that's what VS MSBuild ships)

Follow-up PR:
* Add a variant to the debugger tests that runs with webcil.

lambdageek committed:
Implement WebCIL loader; on by default

It will try to look for WebCIL formatted images in normal .dll files

lambdageek committed:
Checkpoint works on wasm sample; add design doc

lambdageek committed:
Remove unused fields

lambdageek committed:
fixup markdown

lambdageek committed:
remove unused var

lambdageek committed:
specify FileAccess

lambdageek committed:
FileShare

lambdageek commented:
Current status: it's working for basic samples, but failing with the libraries tests, for example, because various bits of code above the runtime assume that assemblies have .dll extensions.  

I was hoping to avoid "logical name"/"physical name" shenanigans, but I might need to adjust the wasm driver.c to consume .webcil files from the app bundle but expose them to the runtime as if they were ".dll".  I'm going to press on without that for a bit longer and try to adjust the loader's search algorithm to just substitute .webcil when people ask for .dll, and if it doesn't pan out, I'll change the bundling logic instead.

jkoritzinsky commented:
Would this be an interesting format to support on all platforms? Or just WebAssembly platforms? Should we define tooling to be able to emit to this format directly (a la APIs in System.Reflection.Metadata)?

Also cc: @AaronRobinsonMSFT who is doing some work with metadata parsing in unmanaged code.

lambdageek commented:
> Would this be an interesting format to support on all platforms? Or just WebAssembly platforms? Should we define tooling to be able to emit to this format directly (a la APIs in System.Reflection.Metadata)?
> 
> Also cc: @AaronRobinsonMSFT who is doing some work with metadata parsing in unmanaged code.

Yea I was going to noodle around with making something like `System.Reflection.Metadata.WebCIL` similar to `SRM.PortableExecutable` over the holidays.  (It's not something I pursued from the start because SRM is [not well suited for the kind of roundtripping that we're doing in the WasmAppBuilder task](https://github.com/dotnet/runtime/issues/30443))

But to be honest, I'm somewhat undecided if this is just an internal implementation detail of the final publishing step for WASM or if it's something that we want to use more broadly.  For example, we may not want to keep a strict 1-1 correlation between the assemblies that comprise an app and the .webcil files that comprise an app bundle - it would be cool to split CoreLib and the main app assembly into hot/cold parts that can be downloaded separately, for example.  Also I don't think we're entirely settled on whether we will serve .webcil directly or if it will be further bundled (for example by baking webcil data into web assembly .wasm modules).

lambdageek committed:
Push .dll->.webcil probing lower in the bundle logic

lambdageek committed:
Also convert satellite assemblies

lambdageek committed:
satellite matching

lambdageek committed:
fixup asset entry

lambdageek committed:
[wasm] don't leak .webcil image names to the debugger

In particular this will make source and breakpoint URLs look like
`dotnet://foo.dll/Foo.cs` which means that grabbing PDBs via source
link will work, etc.

lambdageek committed:
fix one other PE file check

lambdageek commented:
The debugger tests are failing because `mono_wasm_add_assembly` is calling `mono_has_pdb_checksum` which is supposed to look for a PDB checksum in a `.dll`, but that code doesn't use the loader plugin mechanism, so it doesn't know anything about .webcil files.  And also .webcil files don't initialize `MonoDotNetHeader:datadir.pe_debug` properly.

lambdageek committed:
Add PE DebugTableDirectory to webcil

This is used to retrieve the PPDB data and/or the PDB checksum from an
image.

Refactor mono_has_pdb_checksum to support webcil in addition to PE images

lambdageek committed:
fix windows build

lambdageek committed:
Implement a WebcilReader for BorwserDebugProxy like PEReader

This needs some improvements:
- add support for reading CodeView and EmbeddedPDB data
- copy/paste less from the WebcilWriter task
- copy/paste less from PEReader (will require moving WebcilReader to SRM)

lambdageek commented:
Ok, next roadblock is that we actually need a `WebcilReader` (analogous to `PEReader`) for the browser debug proxy - it reads bytes received in the browser and tries to figure out how to get PDB data or CodeView data out of them.  Working on this now...

lambdageek committed:
REVERT ME: temporarily don't try to read codeview data for webcil

lambdageek committed:
[debug] Match bundled pdbs if we're looking up .webcil files

The pdbs are registered by wasm with a notional .dll filename.
if the debugger does a lookup using a .webcil name instead, allow the match

lambdageek committed:
be less verbose

lambdageek commented:
@buyaa-n The problem is we need the `[StructLayout(LayoutKind.Sequential)]` attribtues on `RuntimeAssemblyBuilder`, `RuntimeModuleBuilder`, etc, in the mono versions.  Otherwise the classes are laid out using Auto layout, which rearranges managed object and valuetype fields.  That breaks the runtime, because the reflection objects are accessed from native code using the mirror structures in `object-internals.h`.

lambdageek commented:
Next set of failures is because the PE/COFF debug directory entries have a file pointer (ie a raw file offset) to locate the actual embedded PPDB or CodeView data.  So I guess the WebcilWriter will need to fix those up when writing them out 🤮 

Or maybe I can fudge the file format to have a header the same size as the original PE/COFF file and just fill the intervening space with zeroes?  Not sure which is more of a hack

lambdageek committed:
Adjust debug directory entries when writing webcil files

the PE COFF debug directory entries contain a 'pointer' field which is an
offset from the start of the file.

When writing the webcil file, the header is typically smaller than a
PE file, so the offsets are wrong.  Adjust the offsets by the size of
the file.

We assume (and assert) the debug directory entries actually point at some PE COFF
sections in the PE file (as opposed to somewhere past the end of the
known PE data).

When writing, we initially just copy all the sections directly, then
seek to where the debug directory entries are, and overwrite them with
updated entries that have the correct 'pointer'

lambdageek committed:
Fix bug in WebcilWriter

Stream.CopyTo takes a buffer size, not the number of bytes to copy.

lambdageek committed:
bugfix: the debug directory is at pe_debug_rva not at the CLI header

lambdageek committed:
skip debug fixups if there's no debug directory

lambdageek committed:
WebcilReader: implement CodeView and Emebedded PPDB support

lambdageek commented:
Debugger tests are passing locally; some WBT tests are failing on CI, not sure why yet.

lambdageek committed:
[WBT] Add UseWebcil option (default to true)

lambdageek committed:
rename WebcilWriter -> WebcilConverter [NFC]

lambdageek committed:
fixup AssemblyLoadedEventTest

lambdageek committed:
hack: no extension on assembly for breakpoint

lambdageek committed:
pass normal .dll name for MainAssemblyName in config

let the runtime deal with it - bundle matching will resolve it to the .webcil file

lambdageek commented:
/azp run runtime-wasm

lambdageek commented:
Not sure what's up with `Build Browser wasm Linux Release Mono_DebuggerTests_chrome`.  When I run it locally (on osx-arm64) those tests work.

lambdageek committed:
Wasm.Debugger.Tests: give CI 10 more minutes

lambdageek commented:
@dotnet/area-system-reflection-metadata Hi, SRM folks.  I'd like to make a little helper library ("Microsoft.WebAssembly.Webcil" or some such name) for working with .webcil images and it would be really convenient if I could add `[InternalsVisibleTo("Microsoft.WebAssembly.Webcil")]` to System.Reflection.Metadata.   Specifically, I'd really like to use `AbstractMemoryBlock`  (and its concrete implementations) as well as some of the `PortableExecutable` internals.  Would that be reasonable?

(Webcil is a new container format I'm defining in this PR that we would use to publish .NET apps targeting WebAssembly. The file format is not intended to be user-inspectable, but our tooling (for example debugger adapter that sits between VS and Chrome) needs to understand the files somewhat.)

Right now I've had to resort to copy/pasting and a couple of dodgy uses of reflection to call some internal methods.

lambdageek commented:
Actually, will `InternalsVisibleTo` really solve my problems?  At least in one place `Microsoft.WebAssembly.Webcil` would need to be a `netstandard2.0` or `net462` library because we'd use it from the VS version of MSBuild.  So I'd be depending on the version of SRM that ships with VS - which won't have my IVT.

I might need to copy/paste what I need from SRC (or include via source includes) into the net462 version of Webcil...

lambdageek committed:
Add Microsoft.NET.WebAssembly.Webcil assembly project

Mark it as shipping, but not shipping a nuget package.

The idea is that it will be shipped along with the WasmAppBuilder
msbuild task, and with the BrowserDebugProxy tool.

lambdageek committed:
Move WebcilConverter to Microsoft.NET.WebAssembly.Webcil

lambdageek committed:
Move WebcilReader to Microsoft.NET.WebAssembly.Webcil

delete the duplicated utility classes

lambdageek commented:
I think this is ready for review although there are still a couple of things that need to be done:

1. Turn off Mono support for webcil on non-WebAssembly platforms
2. Turn webcil bundling off by default.  Add an MSBuild property that users can set to preview webcil bundling.
3. Add a couple of Wasm.Build.Test tests that explicitly turns on webcil  (wasmbrowser and blazor templates)
4. Either run debugger tests with webcil by default, or else add a variant to the debugger tests that runs with webcil.   

/cc @lewing

lambdageek commented:
@radical @thaystg @vargaz Could you take a look please

vargaz commented on docs/design/mono/webcil.md:
````diff
@@ -0,0 +1,110 @@
+
+``` c
+struct WebcilHeader {
+	uint8_t id[2]; // 'W' 'C'
````
The magic header should be longer to allow tools like 'file' to identify this file format.
The version should probably be 4 bytes as well (major-minor).

lambdageek replied:
yea I can make the magic longer.

Do we really need a major/minor version scheme? I can make it a bit wider, but I don't think we should be very flexible about mismatches. It's like the AOT image format - you have to match the runtime's expected version exactly, or else we fail.

vargaz replied:
Maybe not, but it's better to have it because it would be hard to add it later.

lambdageek replied:
Ok, changed the magic to `WbIL` and made the version two 16 bit values.

lambdageek committed:
make the webcil magic and version longer

radical commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs:
```diff
@@ -0,0 +1,332 @@
+            long pos = TranslateRVA((uint)_corHeaderMetadataDirectory.RelativeVirtualAddress);
+            if (_stream.Seek(pos, SeekOrigin.Begin) != pos)
+            {
+                throw new BadImageFormatException("Could not seek to metadata", nameof(_stream));
```
Is `nameof(_stream)` useful here? The caller might have the relevant filename.

lambdageek replied:
`nameof(_stream)` is not useful.  But it turns out we don't have any name we can show at all - the DebugStore calls the WebcilReader and it only has an array of bytes from the runtime - until it tries to read the metadata it has no idea what the name of the assembly is.  I'll fix this not to use `nameof` here, but turns out we don't have anything we can show instead.

radical commented on src/mono/mono/metadata/image.c:
```diff
@@ -923,26 +928,44 @@ do_load_header (MonoImage *image, MonoDotNetHeader *header, int offset)
+	int32_t ret = do_load_header_internal (raw_data, raw_data_len, cli_header, offset, FALSE);
+	return ret;
+}
+
```
```suggestion
```

radical commented on src/mono/mono/metadata/mono-debug.c:
```diff
@@ -1097,13 +1097,31 @@ mono_register_symfile_for_assembly (const char *assembly_name, const mono_byte *
+	if (p && *(p + 7) == 0) {
+		size_t n = p - assembly_name;
+		if (!strncmp (bsymfile->aname, assembly_name, n)
+		    && !strcmp (bsymfile->aname + n, ".dll"))
```
```suggestion
			&& !strcmp (bsymfile->aname + n, ".dll"))
```

radical commented on src/mono/mono/metadata/reflection.c:
```diff
@@ -1285,7 +1285,7 @@ method_body_object_construct (MonoClass *unused_class, MonoMethod *method, gpoin
 		(method->flags & METHOD_ATTRIBUTE_ABSTRACT) ||
 	    (method->iflags & METHOD_IMPL_ATTRIBUTE_INTERNAL_CALL) ||
-	    (image->raw_data && image->raw_data [1] != 'Z') ||
+	    (image->raw_data && (image->raw_data [1] != 'Z' && image->raw_data [1] != 'b')) ||
```
```suggestion
		(image->raw_data && (image->raw_data [1] != 'Z' && image->raw_data [1] != 'b')) ||
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -867,36 +868,54 @@ internal sealed class AssemblyInfo
+            {
+                // This is a WebAssembly file
+                asmStream.Seek(0, SeekOrigin.Begin);
+                var webcilReader = new WebcilReader(asmStream);
```
```suggestion
                using var webcilReader = new WebcilReader(asmStream);
```

radical reviewed (commented):
Posting for some of the parts that I looked at. I'll looking at the build, and tests side of this later, and then post any feedback.

lambdageek committed:
fixups

lambdageek commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/Directory.Build.props:
```diff
@@ -0,0 +1,11 @@
+<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
+      <IsShipping>true</IsShipping>
```
It's an [arcade SDK property](https://github.com/dotnet/arcade/blob/5730ab692bd256e6005bd7a03368cd2c219e1954/Documentation/ArcadeSdk.md?plain=1#L907).  AFAICT, You have to set `IsShipping` to true and `IsShippingPackage` to false so that the build logic will do all the normal signing, version info, etc, but skip making a nuget.

lambdageek commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Microsoft.NET.WebAssembly.Webcil.csproj:
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>$(NetCoreAppToolCurrent);$(NetFrameworkToolCurrent)</TargetFrameworks>
```
This library is also used by the `WasmAppBuilder` task, which runs on the .NET Framework MSBuild when started by VS.

lambdageek commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Microsoft.NET.WebAssembly.Webcil.csproj:
```diff
@@ -0,0 +1,28 @@
+    <Compile Include="Webcil\**\*.cs" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
+    <Compile Include="Common\IsExternalInit.cs" />
+  </ItemGroup>
```
I don't think stuff in src/libraries has EnableDefaultCompileItems=true.  All the other libraries explicitly list their sources.

lambdageek commented on src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilConverter.cs:
```diff
@@ -0,0 +1,341 @@
+
+    public void ConvertToWebcil()
+    {
+        using var inputStream = File.Open(_inputPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
```
`FileShare`  is setting what other Streams (incl in other processes, AFAICT) can do with this file.  ReadWrite is the most permissive.  When I had something else here, I had build failures because I think MSBuild and the WasmAppBuilder task were all trying to access the same input file for various reasons.

It does seem suspicious that other streams could be writing to the same file while we try to read it, but it might just be that the other access is done but just didn't close the stream yet. Something like that.

lambdageek committed:
Code style improvements from review

Co-authored-by: Ankit Jain <radical@gmail.com>

lambdageek committed:
Improve some exception messages, when possible

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/BrowserDebugProxy.csproj:
```diff
@@ -14,4 +14,8 @@
   </ItemGroup>
 
+  <ItemGroup>
+      <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.NET.WebAssembly.Webcil\src\Microsoft.NET.WebAssembly.Webcil.csproj" Private="true" />
```
nit: use 4 spaces and not 6 spaces.

thaystg commented on src/mono/wasm/debugger/DebuggerTestSuite/MonoJsTests.cs:
```diff
@@ -136,7 +136,7 @@ async Task AssemblyLoadedEventTest(string asm_name, string asm_path, string pdb_
             });
 
-            byte[] bytes = File.ReadAllBytes(asm_path);
+            byte[] bytes = File.Exists(asm_path) ? File.ReadAllBytes(asm_path) : File.ReadAllBytes(Path.ChangeExtension(asm_path, ".webcil")); // hack!
```
Maybe we need to take look at this code?
https://github.com/dotnet/aspnetcore/blob/main/src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs
Not sure if the Blazor Lazy Assembly Loader will continue working with the dll's using .webcil extension.

lambdageek replied:
I thought the workload build tests would try it.  but I can try manually, too.

lambdageek replied:
Hmm... ah ok, lazy loading is pretty complicated actually. https://learn.microsoft.com/en-us/aspnet/core/blazor/webassembly-lazy-load-assemblies?view=aspnetcore-7.0
I'll try the example from the documentation. you're probably right it needs some  changes in the blazor service.

Also, we might not need this working in the first preview.

thaystg reviewed (commented):
I think you also need to add the new dll, Microsoft.NET.WebAssembly.Webcil.dll, used by BrowserDebugHost in the transport package used by Blazor here: https://github.com/dotnet/runtime/blob/main/src/mono/nuget/Microsoft.NETCore.BrowserDebugHost.Transport/Microsoft.NETCore.BrowserDebugHost.Transport.pkgproj

lambdageek committed:
Suggestings from code review

lambdageek committed:
Add WasmEnableWebcil msbuild property.  Off by default

lambdageek committed:
Build non-wasm runtimes without .webcil support

lambdageek committed:
XXX HACK - Remove - turn on webcil by default on browser-wasm

lambdageek committed:
Run WBT twice: with and without webcil

This is a total of 4 runs: with and without workloads x with and
without webcil

lambdageek committed:
do the cartesian product correctly in msbuild

lambdageek committed:
Revert "XXX HACK - Remove - turn on webcil by default on browser-wasm"

This reverts commit 0f14c3b7c613ad5569316a757587b10dbe2aaa02.

lambdageek committed:
also add webcil to template projects

lambdageek committed:
environment variable has to be non-null and "true"

We set it to "false" sometimes

lambdageek committed:
Fix wasm work items

They shoudl be the same whether or not webcil is used.  Just the
WorkloadItemPrefix should be used to change the name.

Co-Authored-By: Ankit Jain <radical@gmail.com>

radical commented on src/libraries/sendtohelix-wasm.targets:
```diff
@@ -260,7 +261,7 @@
       </HelixWorkItem>
 
-      <HelixWorkItem Include="NoWorkload-Wasm.Build.Tests" Condition="'$(TestUsingWorkloads)' != 'true'">
+      <HelixWorkItem Include="$(WorkloadItemPrefix)Wasm.Build.Tests" Condition="'$(TestUsingWorkloads)' != 'true'">
```
```suggestion
      <HelixWorkItem Include="$(WorkItemPrefix)Wasm.Build.Tests" Condition="'$(TestUsingWorkloads)' != 'true'">
```

lambdageek committed:
Update src/libraries/sendtohelix-wasm.targets

Co-authored-by: Ankit Jain <radical@gmail.com>

lambdageek committed:
FIXME: why is this crashing in MetadataLoadContext

lambdageek committed:
PInvokeTableGeneratorTests: don't try to use the net472 WasmAppBuilder

Look for the default target framework subdirectory under the tasks
directory in the runtime pack when trying to find the tasks dll.
In particular don't try to load the net472 version on modern .NET

lambdageek committed:
PInvokeTableGeneratorTests: Add more diagnostic output if tasksDir is not found

lambdageek commented:
Added an issue to track follow-up work https://github.com/dotnet/runtime/issues/80807

lambdageek committed:
nullability

pavelsavara commented on src/mono/mono/metadata/assembly.c:
```diff
@@ -1449,6 +1450,26 @@ absolute_dir (const gchar *filename)
+	if (g_str_has_suffix (bundled_name, ".webcil") && g_str_has_suffix (name, ".dll")) {
+		size_t bprefix = strlen (bundled_name) - 7;
+		size_t nprefix = strlen (name) - 4;
+		size_t longer = MAX(bprefix, nprefix);
```
you can return false if one of the names is longer

radical commented on src/tasks/WasmAppBuilder/WebcilConverter.cs:
```diff
@@ -0,0 +1,44 @@
+
+    public void ConvertToWebcil()
+    {
+        Log.LogMessage($"Converting to Webcil: input {_inputPath} output: {_outputPath}");
```
Do we want this to be visible to the user in the normal case? If not, we should use `Log.LogMessage(MessageImportance.Log, $"..`

radical replied:
And we could add a user visible message somewhere in WasmAppBuilder saying `Converting assemblies to WebCil`.

Feedback can be addressed, if you choose to, in a follow up PR.

radical reviewed (approved):
Great work! 👍

lambdageek committed:
simplify prefix comparison in bundled_assembly_match

lambdageek committed:
WasmAppBuilder improve logging

Just emit a single Normal importance message about webcil; details as Low importance.

lambdageek commented:
/azp run runtime

lambdageek commented:
/azp run runtime-wasm

lewing commented:
does this have any conflicts with https://github.com/dotnet/runtime/pull/79284

thaystg commented:
> does this have any conflicts with #79284

Merge conflicts probably yes, theoretical conflicts I don't think so, once we still have information that we need to get the symbols from symbol server, like pdbGuid, it's just in another format. We also use the real dll name to find it on symbol server, not sure if this will not be `myDll.dll` and it will become `myDll.webcil` if webcil is enabled, then we will probably have an issue here.

lewing commented on src/mono/mono/metadata/mono-debug.c:
```diff
@@ -1097,13 +1097,31 @@ mono_register_symfile_for_assembly (const char *assembly_name, const mono_byte *
+#ifdef ENABLE_WEBCIL
+	const char *p = strstr (assembly_name, ".webcil");
+	/* if assembly_name ends with .webcil, check if aname matches, with a .dll extension instead */
+	if (p && *(p + 7) == 0) {
```
```suggestion
	if (p && *(p + strlen(".webcil") == 0) {
```
the compiler will do the right thing

lewing commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -187,9 +188,22 @@ private bool ExecuteInternal ()
-            FileCopyChecked(assembly, Path.Combine(asmRootPath, Path.GetFileName(assembly)), "Assemblies");
+            if (UseWebcil)
+            {
+                var tmpWebcil = Path.GetTempFileName();
+                var webcilWriter = Microsoft.WebAssembly.Build.Tasks.WebcilConverter.FromPortableExecutable(inputPath: assembly, outputPath: tmpWebcil, logger: Log);
+                webcilWriter.ConvertToWebcil();
+                var finalWebcil = Path.ChangeExtension(assembly, ".webcil");
+                FileCopyChecked(tmpWebcil, Path.Combine(asmRootPath, Path.GetFileName(finalWebcil)), "Assemblies");
```
is there a reason you aren't doing something like
```suggestion
                var finalWebcil = Path.Combine(asmRootPath, Path.GetFileName(Path.ChangeExtension(assembly, ".webcil")), "Assemblies");
                var webcilWriter = Microsoft.WebAssembly.Build.Tasks.WebcilConverter.FromPortableExecutable(inputPath: assembly, outputPath: finalWebcil, logger: Log);
                webcilWriter.ConvertToWebcil();
                _fileWrites.Add(finalWebcil);
```

lambdageek replied:
@lewing I think I misunderstood `FileCopyChecked`.  I thought it was avoiding writing to `dst` if it already exists with the same content.  My intention was to avoid messing with modification times of `finalWebcil` unless something really changed.

radical replied:
You can use `Utils.CopyIfDifferent(string src, string dst, bool useHash)`.

radical replied:
https://github.com/dotnet/runtime/blob/0e12347861940389153ee2233ff398d551f86711/src/tasks/WasmAppBuilder/EmccCompile.cs#L216-L219

lewing committed:
Add missing using

lewing commented:
we'll need to update https://github.com/dotnet/sdk/blob/ea03dc7f05bfbaac843dc2ac09e26a484acfe8f8/src/BlazorWasmSdk/Targets/BlazorWasm.web.config#L13 as well

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
XXX REVERT ME - temporarily disable dedup

lambdageek commented:
temporarily disabling dedup and re-running

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
Revert "XXX REVERT ME - temporarily disable dedup"

This reverts commit 5983e65d3ba9b638a0aeb08cff810d406d9c3d69.
